### PR TITLE
Update - add option to filter reader by unseen posts

### PR DESCRIPTION
**Pull request by blackjackkent**
#### Changes proposed in this Pull Request

This PR takes a stab at adding a toggle-able control on Reader list pages to allow the user to hide all posts which they have already seen.

#### Testing instructions

* Navigate to `/read`
* Verify that an "Unread" menu option shows in the sidebar under "Following" and "A8C".
* Navigate to these views and verify that already-read posts appear and disappear as expected.

image
image

Related to (no corresponding issue - 20% time task)

### CODE REVIEW

**Comment by kwight** [comment]
I wonder if "Hide unread/show unread" would be clearer? I'll tag design too to see if there's any existing pattern for this sort of button/filter. And being a filter, I wonder if a segmented control style would be more appropriate (like in the notifications)?

Screen Shot 2021-03-22 at 10 15 00 AM

Can the button's state be persisted? I noticed the Reader view always loads with it off. If I activate the button, click on a post to read it, then hit Back, it'll load in the "all posts" state (when I expected it to be the same state as I left it in). It also strikes me as a "set once and forget" type setting.

**Comment by sixhours** [comment]
This does feel like filter or sort behavior, using the segmented control like Kirk suggests. Clicking the "Unread" filter could bubble all unread posts to the top, and they'd automatically drop down below the unread list after they've been read. Or maybe we introduce a view in the left nav that only shows unread posts.

Putting on my designer hat, is the main problem that the screen is too crowded? That it's hard to find unread content among the rest?

**Comment by blackjackkent** [comment]
> This does feel like filter or sort behavior, using the segmented control like Kirk suggests. Clicking the "Unread" filter could bubble all unread posts to the top, and they'd automatically drop down below the unread list after they've been read. Or maybe we introduce a view in the left nav that only shows unread posts.
> 
> Putting on my designer hat, is the main problem that the screen is too crowded? That it's hard to find unread content among the rest?

I think for me personally, it's just that the way my brain works makes it satisfying to work through the unread stuff and have it clear off the screen when I've looked at it (for instance when I'm getting caught up on P2s). I think it's more a matter of personal preference than the current design being bad per se, which is why this tweak is intended to give options.

I do like the idea too of having a specific "Unread Posts" view; I take it you think that's preferable to having it be a selectable control?

**Comment by sixhours** [comment]
> I think for me personally, it's just that the way my brain works makes it satisfying to work through the unread stuff and have it clear off the screen when I've looked at it (for instance when I'm getting caught up on P2s).

I definitely get that. It does feel odd to me to potentially have a ton of posts grayed out at the top of the list, pushing down unread content below. And I like the satisfaction of "inbox zero", so hiding the stuff I've already seen makes sense.

> I do like the idea too of having a specific "Unread Posts" view; I take it you think that's preferable to having it be a selectable control?

I think a control could work, but having three CTAs to choose from in that top bar is throwing me off. Looking at the buttons side by side, I'd expect one primary call to action and one secondary (or maybe tertiary) call to action.

My thought process is, since it feels like a filter, using a filter component (something like the notifications filter bar or the tabbed filters we use for posts and pages -- Published, Draft, Private, Trash, etc.) makes more sense. Which is how I ended up at "Oohh, it would nice to have a view that only shows unread posts" :D

**Comment by sixhours** [comment]
Not sure if you're ready for this level of granular feedback, but I took it for a spin and I got some weird results when I first click "Mark all as seen" and then "Show unread only":

Screen Shot 2021-03-29 at 4 07 57 PM

I also noticed there are unread posts when Infinite Scroll kicks in and shows more, but I can't click "Mark all as seen" again. But that's happening on trunk too, so. 🤷‍♀️

**Comment by blackjackkent** [comment]
Per @sixhours comments, I have updated this to set up "Unread" view menu items rather than having it as a clickable button. I'm updating the testing instructions.

**Comment by kwight** [comment]
> I also noticed there are unread posts when Infinite Scroll kicks in and shows more, but I can't click "Mark all as seen" again. But that's happening on trunk too, so. 🤷‍♀️

What I noticed is that the blogs themselves are being listed, but with each showing "0 posts" (since, well, they've been read). Could an empty "Unread" view show nothing at all, or better yet, maybe a "You're all caught up!" message, like notifications? Having anything in an empty Unread view seems odd to me. 🤔 

Screen Shot 2021-04-08 at 12 29 34 PM

File: client/reader/controller.js [hunk 1]
```diff
@@ -349,3 +352,7 @@ export const {
 	readA8C,
 	readFollowingP2,
 } = exported;
+
+export const unreadFollowing = ( context, next ) => exported.following( context, next, false );
+export const unreadA8C = ( context, next ) => exported.readA8C( context, next, false );
+export const unreadFollowingP2 = ( context, next ) => exported.following( context, next, false );
```
Comment (mreishus) [inline client/reader/controller.js]: should this use `exported.readFollowingP2()`? right now `unreadFollowing` and `unreadFollowingP2` do the same thing

**Comment by mreishus** [comment]
I'm also seeing "0 posts" entries. I do have unread posts, but my read posts turn into site name + 0 posts when I switch to unread view.  Screenshots:

### All View


### Unread View

**Comment by blackjackkent** [comment]
I pushed an update for this which should filter out the combined cards from the Reader when they have no posts left to display after filtering out read ones. @mreishus @sixhours can you confirm?

**Comment by mreishus** [comment]
Confirmed :+1: Those site headings containing only unread posts are correctly filtered out now.

File: client/reader/sidebar/reader-sidebar-organizations/list.jsx [hunk 2]
```diff
@@ -71,6 +71,33 @@ export class ReaderSidebarOrganizationsList extends Component {
...
 
+	renderUnread() {
+		const { translate, organization, path, sites } = this.props;
+		// have a selector
+		const sum = sites.reduce( ( acc, item ) => {
+			acc = acc + item.unseen_count;
+			return acc;
+		}, 0 );
+		return (
+			<>
```
Comment (kwight) [inline client/reader/sidebar/reader-sidebar-organizations/list.jsx]: With only one child, we shouldn't need a fragment here, I don't think.

**Comment by kwight** [comment]
This is super close! I'm not seeing any post content when Unread is empty. But I am seeing an infinite scroll of Recommended Posts – they keep loading if I keep scrolling, and then the app crashes after also having some previous post content at the end.

Maybe the Unread view could just have a "you're all caught up" type message, followed by one row of Recommended Posts, and that's it?

Screen Shot 2021-04-20 at 7 50 04 AM

**Comment by kwight** [comment]
@blackjackkent What would you like to do with this? I'm triaging our old-is-new Serenity board. :)

**Comment by kwight** [comment]
@blackjackkent I've removed this from our active board – whenever you get time to follow up and improve it, feel free to add it back! 👍

**Comment by blackjackkent** [comment]
Finally made my way back to this and have updated it to include a "You're all caught up!" image/message when the user has no posts to display in an "Unread" view. (Incorporating the Recommended blocks into that use case ended up being a bit hairy compared to just doing it this way, but I can go back and work further on it if it seems like a blocker.)

image

I've also rebased this against all the changes that happened in trunk since I was last here so this should be pretty much ready for primetime pending any further reviewer comments.

**Comment by sixhours** [comment]
I started testing this, and in the process of clicking around I managed to remove/hide my followed sites list:
Screen Shot 2022-02-15 at 1 44 44 PM

I think I might have clicked "Mark all as seen", but I don't remember exactly. 😬 

* I'm seeing the same thing on production, so I wonder if something in this PR cleared the state?
* I can still see posts from my followed sites, but the sites themselves don't appear in the sidebar with unread counts. 
* I checked another test account and I still see the sites in the sidebar.
* I cleared browser cache/cookies; no change.

@blackjackkent Are you able to reproduce this on trunk and/or with this branch?

File: client/reader/controller.js (hunk 1 above)
Comment (kwight) [reply client/reader/controller.js]: Does this still apply @blackjackkent ?

**Comment by kwight** [comment]
I still get a pile (10+) of Recommended Posts before any actual unread content. 🤔 I noticed the A8C group doesn't have the Unread section, but it seems the dropdown is missing in prod too.

Screen Shot 2022-03-02 at 9 06 17 AM

**Comment by markbiek** [comment]
Everything seemed to work fine for me on "Followed A8C Sites" but I'm getting an unexpected error when I try "Followed Sites".

**Steps to reproduce**

1. Starting from http://calypso.localhost:3000/read/a8c/unread
2. Expanded "Followed Sites"
image
3. Clicked "Unread" and immediately get "We're sorry, but an unexpected error has occurred".

The dev console shows this
image

I stepped through that function and, even though there are five items in the `posts` array, two of them are `undefined`.
image

If it helps, all of the unread posts are from my markbiek132375254.wpcomstaging.com test site (which only has three posts in the admin).

**Comment by blackjackkent** [comment]
Closing this for reasons described in paYKcK-2gO-p2 - will open a fresh PR if I decide to look at this again in the future.

**Pull request closed without merging**